plugins/woocommerce-admin/client/products/product-page.tsx

```diff
@@ -1,19 +1,73 @@
 /**
  * External dependencies
  */
-import { __experimentalEditor as Editor } from '@woocommerce/product-editor';
+import {
+	__experimentalEditor as Editor,
+	AUTO_DRAFT_NAME,
+} from '@woocommerce/product-editor';
 import { Product } from '@woocommerce/data';
+import { useDispatch, useSelect, select as WPSelect } from '@wordpress/data';
+import { useEffect, useState } from '@wordpress/element';
+import { Spinner } from '@wordpress/components';
+import { useParams } from 'react-router-dom';
 
 /**
  * Internal dependencies
  */
 import './product-page.scss';
 
-const dummyProduct = {
-	name: 'Example product',
-	short_description: 'Short product description content',
-} as Product;
+const ProductEditor: React.FC< { product: Product | undefined } > = ( {
+	product,
+} ) => {
+	if ( ! product ) {
+		return <Spinner />;
+	}
+
+	return <Editor product={ product } settings={ {} } />;
+};
+
+const EditProductEditor: React.FC< { productId: string } > = ( {
+	productId,
+} ) => {
+	const { product } = useSelect( ( select: typeof WPSelect ) => {
+		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+		// @ts-ignore Missing types.
+		const { getEditedEntityRecord } = select( 'core' );
+
+		return {
+			product: getEditedEntityRecord( 'postType', 'product', productId ),
+		};
+	} );
+
+	return <ProductEditor product={ product } />;
+};
+
+const AddProductEditor = () => {
+	const { saveEntityRecord } = useDispatch( 'core' );
+	const [ product, setProduct ] = useState< Product | undefined >(
+		undefined
+	);
+
+	useEffect( () => {
+		saveEntityRecord( 'postType', 'product', {
+			title: AUTO_DRAFT_NAME,
+			status: 'auto-draft',
+			// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+			// @ts-ignore Incorrect types.
+		} ).then( ( autoDraftProduct: Product ) => {
+			setProduct( autoDraftProduct );
+		} );
+	}, [] );
+
+	return <ProductEditor product={ product } />;
+};
 
 export default function ProductPage() {
-	return <Editor product={ dummyProduct } settings={ {} } />;
+	const { productId } = useParams();
+
+	if ( productId ) {
+		return <EditProductEditor productId={ productId } />;
+	}
```

Comment:
Instead of separating this logic into two separate (ultimately 3) components would it be worth just moving the product logic into a custom hook, that either returns a new auto draft product or an existing product?

That way you can just call `const product = useProductEntityRecord( productId )` and pass that straight into the Editor.
Only suggesting this, cause otherwise I think we are better off splitting this into separate files again (but we were actually trying to avoid that).